src/txn/txn.c

```diff
@@ -1095,25 +1094,27 @@ __txn_commit_timestamps_assert(WT_SESSION_IMPL *session)
           (upd->txnid == WT_TXN_ABORTED || upd->txnid == WT_TXN_NONE || upd->txnid == txn->id))
             upd = upd->next;
 
+        if (upd == NULL)
+            continue;
+
         /*
          * Check the timestamp on this update with the first valid update in the chain. They're in
          * most recent order.
          */
-        if (upd != NULL) {
-            prev_op_timestamp = upd->start_ts;
-            durable_op_timestamp = upd->durable_ts;
-        }
+        prev_op_timestamp = upd->start_ts;
+        durable_op_timestamp = upd->durable_ts;
 
-        if (upd == NULL)
-            continue;
         /*
          * Check for consistent per-key timestamp usage. If timestamps are or are not used
          * originally then they should be used the same way always. For this transaction, timestamps
          * are in use anytime the commit timestamp is set. Check timestamps are used in order.
+         *
+         * We may see an update restored from the history store with 0 timestamp if that update is
+         * behind the oldest timestamp when the history store page is reconciled.
          */
         op_zero_ts = !F_ISSET(txn, WT_TXN_HAS_TS_COMMIT);
         upd_zero_ts = prev_op_timestamp == WT_TS_NONE;
-        if (op_zero_ts != upd_zero_ts) {
+        if (op_zero_ts != upd_zero_ts && !F_ISSET(upd, WT_UPDATE_RESTORED_FOR_ROLLBACK)) {
```

Comment:
This change makes sense to me. Is rollback the only time we ever create an update structure based on something that has been on disk?

Reply:
We may also append the onpage value to the update chain in reconciliation. But that one is benign in this case as it means we select a newer value to write to disk and there must be a newer update with timestamp on the update chain so the check will pass.